pkg/image/admission/imagepolicy/accept.go

```diff
@@ -26,40 +27,56 @@ type policyDecision struct {
 	err    error
 }
 
-func accept(accepter rules.Accepter, resolver imageResolver, m meta.ImageReferenceMutator, attr admission.Attributes, excludedRules sets.String) error {
+func accept(accepter rules.Accepter, imageResolutionType imagepolicyapi.ImageResolutionType, resolver imageResolver, m meta.ImageReferenceMutator, attr admission.Attributes, excludedRules sets.String) error {
 	var decisions policyDecisions
 
 	gr := attr.GetResource().GroupResource()
-	requiresImage := accepter.RequiresImage(gr)
-	resolvesImage := accepter.ResolvesImage(gr)
 
 	errs := m.Mutate(func(ref *kapi.ObjectReference) error {
 		// create the attribute set for this particular reference, if we have never seen the reference
 		// before
 		decision, ok := decisions[*ref]
 		if !ok {
-			var attrs *rules.ImagePolicyAttributes
-			var err error
-			if requiresImage || resolvesImage {
-				// convert the incoming reference into attributes to pass to the accepter
-				attrs, err = resolver.ResolveObjectReference(ref, attr.GetNamespace())
+			if imagepolicyapi.RequestsResolution(imageResolutionType) {
+				resolvedAttrs, err := resolver.ResolveObjectReference(ref, attr.GetNamespace())
+
+				switch {
+				case err != nil && imagepolicyapi.FailOnResolutionFailure(imageResolutionType):
+					// if we had a resolution error and we're supposed to fail, fail
+					decision.err = err
+					decision.tested = true
```

Comment:
> if you return at bottom, are these lost? I don't see you returning them

added to cache here.  The value is scoped high up, so returning down below (if we tested) will store in the cache.